src/Compiler/Service/SynExpr.fs

```diff
@@ -423,7 +423,9 @@ module SynExpr =
                 | SynExpr.MatchLambda(matchClauses = Last(SynMatchClause(resultExpr = expr)))
                 | SynExpr.MatchBang(clauses = Last(SynMatchClause(resultExpr = expr)))
                 | SynExpr.TryWith(withCases = Last(SynMatchClause(resultExpr = expr)))
-                | SynExpr.TryFinally(finallyExpr = expr) -> loop expr
+                | SynExpr.TryFinally(finallyExpr = expr)
+                | SynExpr.Do(expr = expr)
```

Comment:
Technically, we don't need to continue looping if the `do`-expression ends in `done`, but the presence of that keyword isn't actually visible in the AST except in the `do`-expression's _parent_ nodes' ranges.

<details>
<summary>Compare:</summary>

```fsharp
module M =
    do ()
```

```fsharp
ParsedInput.ImplFile(
    ParsedImplFileInput.ParsedImplFileInput(
        fileName = "tmp.fsx",
        isScript = true,
        qualifiedNameOfFile = QualifiedNameOfFile.QualifiedNameOfFile(Ident("Tmp$fsx", R("(1,0--2,9)"))),
        scopedPragmas = [],
        hashDirectives = [],
        contents = [
            SynModuleOrNamespace.SynModuleOrNamespace(
                longId = [ Ident("Tmp", R("(1,0--1,0)")) ],
                isRecursive = false,
                kind = SynModuleOrNamespaceKind.AnonModule,
                decls = [
                    SynModuleDecl.NestedModule(
                        moduleInfo =
                            SynComponentInfo.SynComponentInfo(
                                attributes = [],
                                typeParams = None,
                                constraints = [],
                                longId = [ Ident("M", R("(1,7--1,8)")) ],
                                xmlDoc = PreXmlDoc.Empty,
                                preferPostfix = false,
                                accessibility = None,
                                range = R("(1,0--1,8)")
                            ),
                        isRecursive = false,
                        decls = [
                            SynModuleDecl.Expr(expr = SynExpr.Do(expr = SynExpr.Const(constant = SynConst.Unit, range = R("(2,7--2,9)")), range = R("(2,4--2,9)")), range = R("(2,4--2,9)"))
                        ],
                        isContinuing = false,
                        range = R("(1,0--2,9)"),
                        trivia = {
                            ModuleKeyword = Some(R("(1,0--1,6)"))
                            EqualsRange = Some(R("(1,9--1,10)"))
                        }
                    )
                ],
                xmlDoc = PreXmlDoc.Empty,
                attribs = [],
                accessibility = None,
                range = R("(1,0--2,9)"),
                trivia = {
                    LeadingKeyword = SynModuleOrNamespaceLeadingKeyword.None
                }
            )
        ],
        flags = (false, false),
        trivia = {
            ConditionalDirectives = []
            CodeComments = []
        },
        identifiers = set []
    )
)
```

against

```fsharp
module M =
    do () done
```

```fsharp
ParsedInput.ImplFile(
    ParsedImplFileInput.ParsedImplFileInput(
        fileName = "tmp.fsx",
        isScript = true,
        qualifiedNameOfFile = QualifiedNameOfFile.QualifiedNameOfFile(Ident("Tmp$fsx", R("(1,0--2,14)"))),
        scopedPragmas = [],
        hashDirectives = [],
        contents = [
            SynModuleOrNamespace.SynModuleOrNamespace(
                longId = [ Ident("Tmp", R("(1,0--1,0)")) ],
                isRecursive = false,
                kind = SynModuleOrNamespaceKind.AnonModule,
                decls = [
                    SynModuleDecl.NestedModule(
                        moduleInfo =
                            SynComponentInfo.SynComponentInfo(
                                attributes = [],
                                typeParams = None,
                                constraints = [],
                                longId = [ Ident("M", R("(1,7--1,8)")) ],
                                xmlDoc = PreXmlDoc.Empty,
                                preferPostfix = false,
                                accessibility = None,
                                range = R("(1,0--1,8)")
                            ),
                        isRecursive = false,
                        decls = [
                            SynModuleDecl.Expr(expr = SynExpr.Do(expr = SynExpr.Const(constant = SynConst.Unit, range = R("(2,7--2,9)")), range = R("(2,4--2,9)")), range = R("(2,4--2,9)"))
                        ],
                        isContinuing = false,
                        range = R("(1,0--2,9)"),
                        trivia = {
                            ModuleKeyword = Some(R("(1,0--1,6)"))
                            EqualsRange = Some(R("(1,9--1,10)"))
                        }
                    )
                ],
                xmlDoc = PreXmlDoc.Empty,
                attribs = [],
                accessibility = None,
                range = R("(1,0--2,14)"),
                trivia = {
                    LeadingKeyword = SynModuleOrNamespaceLeadingKeyword.None
                }
            )
        ],
        flags = (false, false),
        trivia = {
            ConditionalDirectives = []
            CodeComments = []
        },
        identifiers = set []
    )
)
```

</details>

<details>
<summary>It doesn't even show up in the AST at all in some cases:</summary>

```fsharp
module M
do ()
```

```fsharp
ParsedInput.ImplFile(
    ParsedImplFileInput.ParsedImplFileInput(
        fileName = "tmp.fsx",
        isScript = true,
        qualifiedNameOfFile = QualifiedNameOfFile.QualifiedNameOfFile(Ident("M$fsx", R("(1,0--2,5)"))),
        scopedPragmas = [],
        hashDirectives = [],
        contents = [
            SynModuleOrNamespace.SynModuleOrNamespace(
                longId = [ Ident("M", R("(1,7--1,8)")) ],
                isRecursive = false,
                kind = SynModuleOrNamespaceKind.NamedModule,
                decls = [
                    SynModuleDecl.Expr(expr = SynExpr.Do(expr = SynExpr.Const(constant = SynConst.Unit, range = R("(2,3--2,5)")), range = R("(2,0--2,5)")), range = R("(2,0--2,5)"))
                ],
                xmlDoc = PreXmlDoc.Empty,
                attribs = [],
                accessibility = None,
                range = R("(1,0--2,5)"),
                trivia = {
                    LeadingKeyword = SynModuleOrNamespaceLeadingKeyword.Module(R("(1,0--1,6)"))
                }
            )
        ],
        flags = (false, false),
        trivia = {
            ConditionalDirectives = []
            CodeComments = []
        },
        identifiers = set []
    )
)
```

against

```fsharp
module M
do () done
```

```fsharp
ParsedInput.ImplFile(
    ParsedImplFileInput.ParsedImplFileInput(
        fileName = "tmp.fsx",
        isScript = true,
        qualifiedNameOfFile = QualifiedNameOfFile.QualifiedNameOfFile(Ident("M$fsx", R("(1,0--2,5)"))),
        scopedPragmas = [],
        hashDirectives = [],
        contents = [
            SynModuleOrNamespace.SynModuleOrNamespace(
                longId = [ Ident("M", R("(1,7--1,8)")) ],
                isRecursive = false,
                kind = SynModuleOrNamespaceKind.NamedModule,
                decls = [
                    SynModuleDecl.Expr(expr = SynExpr.Do(expr = SynExpr.Const(constant = SynConst.Unit, range = R("(2,3--2,5)")), range = R("(2,0--2,5)")), range = R("(2,0--2,5)"))
                ],
                xmlDoc = PreXmlDoc.Empty,
                attribs = [],
                accessibility = None,
                range = R("(1,0--2,5)"),
                trivia = {
                    LeadingKeyword = SynModuleOrNamespaceLeadingKeyword.Module(R("(1,0--1,6)"))
                }
            )
        ],
        flags = (false, false),
        trivia = {
            ConditionalDirectives = []
            CodeComments = []
        },
        identifiers = set []
    )
)
```

I.e., the AST is identical, even though the source code is not.

</details>

But since the verbose terminator `done` is relatively uncommon anyway, it's very likely preferable that we err on the side of caution here and simply always assume there is no `done`.